### ISSUE TITLE
Parse data URLs (and their MIME types) according to the Fetch Standard

### ISSUE THREAD

**Issue by SimonSapin** [servo/servo#2909]
Spec is at https://fetch.spec.whatwg.org/#data-urls

Status as of November 2019: https://crates.io/crates/data-url has an initial implementation, but integration with Servo is blocked on figuring out the interactions with https://crates.io/crates/mime. CC https://github.com/hyperium/mime/issues/78, https://github.com/servo/rust-url/issues/567.

----

**Initial title:** Figure out the details of data URL parsing

**Initial description:**

For example, what should happen when a base64 sequence contains a (0x20 ASCII) space? Another non-alphabet character?

Since [RFC 2397](http://tools.ietf.org/html/rfc2397) does no define error handling, this might involve reverse-engineering other implementations.

https://www.w3.org/Bugs/Public/show_bug.cgi?id=19494
http://greenbytes.de/tech/tc/datauri/

**Comment by frewsxcv** [comment]
Should this be moved to rust-url?

**Comment by SimonSapin** [comment]
The corresponding code is currently in `components/net/data_loader.rs` in this repository, not in rust-url. I might move that code to rust-url once all the details are figured out and I’m confident to tell the world "this is how you do it" (i.e. when this issue can be closed) but in the mean time I’m not planning to.

**Comment by nox** [comment]
The spec bug trail brought me to https://github.com/whatwg/fetch/issues/234.

**Comment by pshaughn** [comment]
https://infra.spec.whatwg.org/#forgiving-base64-decode might be the thing to do here?

**Comment by jdm** [comment]
Indeed, step 11.2 of https://fetch.spec.whatwg.org/#data-urls uses the forgiving base64 decode algorithm, so that looks like a change we will need to make in our implementation.

**Comment by pshaughn** [comment]
I am doing that now; we're already forgiving spaces, just need to add forgiveness of a few other kinds of byte.

**Comment by pshaughn** [comment]
There is one remaining problem: "ab=" is invalid base64, but we're responding to this particular invalid base64 in a different way from all the other invalid base64. I don't know where this "Reached unreachable code" string is even being generated yet; it doesn't seem to be a string in Servo or in the base64 crate.

`  ▶ Unexpected subtest result in /fetch/data-urls/base64.any.worker.html:
  │ FAIL [expected PASS] data: URL base64 handling: "ab="
  └   → assert_unreached: Should have rejected: undefined Reached unreachable code

  ▶ Unexpected subtest result in /fetch/data-urls/base64.any.html:
  │ FAIL [expected PASS] data: URL base64 handling: "ab="
  └   → assert_unreached: Should have rejected: undefined Reached unreachable code
`

**Comment by jdm** [comment]
`Reached unreachable code` comes from https://github.com/web-platform-tests/wpt/blob/783959beb60fc589e897cc11d08c23f621259f62/resources/testharness.js#L1852 (or https://github.com/servo/servo/blob/f1aa5d8dbdc32f6f474b09234558652b9bead686/tests/wpt/web-platform-tests/resources/testharness.js#L1852 in our vendored tree).

**Comment by SimonSapin** [comment]
If someone is looking at this, please consider using https://crates.io/crates/data-url or filing issues in https://github.com/servo/rust-url/issues about why it’s not ideal :)

**Comment by pshaughn** [comment]
@SimonSapin do you have any advice about how to use the data-url crate in a project that's using https://crates.io/crates/mime elsewhere? I don't see an obvious interconversion strategy between the two classes, even though they expose almost identical getters. I am new to Rust's object model and it's possible there is some obvious answer here.

**Comment by pshaughn** [comment]
One possibility here is to keep the servo code that's parsing some of the URL to make a Mime object from the mime crate, call the data-url crate to get the data, and just ignore the data-url crate's own mime type parsing. That seems like it'd get correct behavior, but it feels a little odd to have to parse the mime type twice and throw one out. Still, I'll try it tomorrow and see if it at least passes all the tests.

**Comment by SimonSapin** [comment]
This is a very relevant question (that was out of my mind when I suggested using `data-url`, and) that unfortunately doesn’t really have an answer yet. See https://github.com/hyperium/mime/issues/78

Parsing twice not only feels odd, but might cause subtle bugs when the two parsers don’t behave the same. So let’s not use the parser in `data-url` yet, keep this issue open, and accept https://github.com/servo/servo/pull/24865 as an incremental improvement.

Would using https://docs.rs/data-url/0.1.0/data_url/forgiving_base64/ (without the rest of the crate) help with the `ab=` case?

**Comment by pshaughn** [comment]
Yes, that saves net/data_loader.rs from having to even know about the definition of ASCII whitespace, and it fixes that one last test case! Thank you!

**Comment by SimonSapin** [comment]
As discussed above, this issue should not be closed yet.

@pshaughn Please note that some keywords in commit messages and issue descriptions are meaningful to GitHub: https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords
